src/App.jsx

```diff
@@ -1,36 +1,172 @@
-import _ from 'lodash';
-import React, { Component } from 'react';
-import './app.scss';
-import { getPets } from './api';
-import PetListItem from './PetListItem.jsx';
+import React, { Component } from "react";
+import "./app.scss";
+import { getPetBatch } from "./api";
+import PetListItem from "./PetListItem.jsx";
+import TableHeader from "./TableHeader.jsx";
+import InfiniteScroll from "react-infinite-scroller";
+import { Button, message, notification } from "antd";
 
 class App extends Component {
   constructor(props) {
     super(props);
     this.state = {
       pets: [],
+      loading: true,
+      hasMore: true,
+      page: 0,
+      sortBy: null,
+      filterBy: null
     };
   }
 
   componentDidMount = async () => {
-    const pets = await getPets();
-    this.setState({ pets });
-  }
+    try {
+      // Load first 20 pets
+      const batch = await getPetBatch(20, 0);
+      const pets = batch.response;
+      this.setState({ pets, page: 1, loading: false });
+    } catch (err) {
+      // Handle error
+      this.openNotification();
+    }
+  };
+
+  handleLoad = async () => {
+    // Prepare to load
+    let pets = this.state.pets;
+    this.setState({
+      loading: true
+    });
+    // Get a new batch
+    try {
+      const newBatch = await getPetBatch(
+        20,
+        this.state.page,
+        this.state.sortBy,
+        this.state.filterBy
+      );
+      // Done displaying
+      if (newBatch.status === false) {
+        message.warning("That's all!");
+        this.setState({
+          hasMore: false,
+          loading: false
+        });
+        return;
+      }
+      pets = [...pets, ...newBatch.response];
+      const newPage = this.state.page + 1;
+      this.setState({ pets, page: newPage, loading: false });
+    } catch (err) {
+      this.openNotification();
+    }
```

Comment:
I'd always set loading to false in a finally block
```
finally {
  this.setState({ loading: false });
}
```

Reply:
Ah, i see you're using the loading flag as a way to prevent the infinite scroller from retrying when an error happens?

There's probably a better design here that doesn't overload `loading` in this way. I'm imagining clicking Try Again just calls `handleLoad` again (`if !loading && hasMore`)